### ISSUE TITLE
Does the tensor in the function used by mode in merge or lambda layers support slice index?

### ISSUE THREAD

**Issue by dingningbuaa** [fchollet/keras#4213]
I would like to implement a network in which there is a cross input difference layer, computing the 4D tensor difference of nearest matrix of every element along axis=1 and axis=2. Here is my attemption, from cross_input function you will know what the layer is, though it raised errors:

from keras import backend as K
from keras.models import Model
from keras.layers import Input,Dense,Convolution2D,Activation,MaxPooling2D,Flatten,merge
from keras.optimizers import SGD,Adam,RMSprop

def cross_input(X):
    tensor_left = X[0]
    tensor_right = X[1]
    #x_length = K.shape(X)[1]
    x_length = 37
    #y_length = K.shape(X)[2]
    y_length = 12
    cross_y = []
    cross_x = []
    tensor_left_padding = K.spatial_2d_padding(tensor_left,padding=(2,2))
    tensor_right_padding = K.spatial_2d_padding(tensor_right,padding=(2,2))
    for i_x in range(2, x_length - 2):
        for i_y in range(2, y_length - 2):
            cross_y.append(tensor_left_padding[:,i_x-2:i_x+3,i_y-2:i_y+3,:] 
                         - tensor_right_padding[:,i_x-2:i_x+3,i_y-2:i_y+3.:])
        cross_x.append(K.concatenate(cross_y,axis=2))
    cross_out = K.concatenate(cross_x,axis=1)
    return K.abs(cross_out)
    
def cross_input_shape(input_shape):
    return (input_shape[0],input_shape[1] * 5, input_shape[2] * 5, input_shape[3])

a1 = Input(shape=(128,64,3))
b1 = Input(shape=(128,64,3))
share = Convolution2D(20,5,5,dim_ordering='tf')
a2 = share(a1)
b2 = share(b1)
a3 = Activation('relu')(a2)
b3 = Activation('relu')(b2)
a4 = MaxPooling2D(dim_ordering='tf')(a3)
b4 = MaxPooling2D(dim_ordering='tf')(b3)
share2 = Convolution2D(25,5,5,dim_ordering='tf')
a5 = share2(a4)
b5 = share2(b4)
a6 = Activation('relu')(a5)
b6 = Activation('relu')(b5)
a7 = MaxPooling2D(dim_ordering='tf')(a6)
b7 = MaxPooling2D(dim_ordering='tf')(b6)
c1 = merge([a7,b7],mode=cross_input,output_shape=cross_input_shape)
c2 = Convolution2D(25,5,5,dim_ordering='tf',activation='relu')(c1)
c3 = MaxPooling2D((2,2),dim_ordering='tf')(c2)
c4 = Convolution2D(25,3,3,dim_ordering='tf',activation='relu')(c3)
c5 = MaxPooling2D((2,2),dim_ordering='tf')(c4)
c6 = Flatten()(c5)
c7 = Dense(10,activation='relu')(c6)
c8 = Dense(2,activation='softmax')(c7)

model = Model(input=[a1,b1],output=c8)

This python script raised several errors, first, it says the x_length in the range(2,x_length-2) should be the integer not tensor, so i set it to a constant. While, it raised another issue about slice index in the code:
cross_y.append(tensor_left_padding[:,i_x-2:i_x+3,i_y-2:i_y+3,:] 
                         - tensor_right_padding[:,i_x-2:i_x+3,i_y-2:i_y+3.:])
I wonder if the tensor in a function to be sent to the merge's mode attribute support the slice index.
If it supports, how to fix this error? If not, how to implement this layer?
